### DISCUSSION TITLE
Possible to build witness from r1CS only?

### DISCUSSION THREAD

**Discussion by Scratch-net** [Consensys/gnark#981] [Q&A] [answered]
Hey there, do I really need the actual circuit code to build witness for proving? Shouldn't R1CS be enough? I'm just thinking of optimizing space for mobile and removing some code would help

**Comment by ivokub** [comment, marked as answer]
You are right, you do not need the full circuit in Go. See for example https://pkg.go.dev/github.com/consensys/gnark@v0.9.1/backend/witness#example-Witness how to define witness for compiled circuit.

**Comment by Scratch-net** [comment]
Actually the opposite is also possible, I can compile circuit each time instead of loading it from a file. Especially that reading r1cs is not working right now :)
